[Alexander Er] iP

aceszhenwei opened the pull request:
## Lazy Duke is the Best Duke

> "I choose a lazy person to do a hard job. Because a lazy person will find an easy way to do it." - Bill Gates [(Source)](https://www.goodreads.com/quotes/568877-i-choose-a-lazy-person-to-do-a-hard-job)

Why you should get Duke:
- Few Commands
- Text Based
- Remembers stuff for you so you can be lazy

All you need is:
1. Download it from [here](https://github.com/aceszhenwei/ip/releases/tag/v1.0)
2. Run it
3. Do stuff
4. ...
5. Profit? 💰 

AND IT IS FREE ~FOREVER~ FOR NOW!

Features:
- [x] Task Management
- [ ] Reminders (coming soon)
- [x] GUI (coming soon)
- [ ] Crypto Mining (we'll get back on this one)

If you want to program too, here's the `main` method.
```java
public static void main(String[] args) {
        new Duke().run();
    }
```
Hopefully that was helpful.

Commit:
Add Gradle support

aceszhenwei committed:
Level-1

aceszhenwei committed:
Level-2

aceszhenwei committed:
Level-3

aceszhenwei committed:
Level-4

Made some adjustments to Task for further abstraction.

aceszhenwei committed:
A-TextUiTesting

Still has issues when comparing, although looks identical. Reverted to Duke name for simplicity.

aceszhenwei committed:
Level-5

aceszhenwei committed:
Level-6

Changed to ArrayLists

aceszhenwei committed:
Fixed Text Testing

aceszhenwei committed:
Level-7

Remade and abstracted out the methods within main.
Utilised a serialized version in order to save the file.

aceszhenwei committed:
Level-8

aceszhenwei committed:
Merge branch 'branch-Level-8'

aceszhenwei committed:
More OOP Part 1

aceszhenwei committed:
More OOP Part 2

Fixed and finished the remainder of More OOP. To merge branches.

aceszhenwei committed:
A-Packages

aceszhenwei committed:
Revert "A-Packages"

This reverts commit 2686fb0a86905581d1578f32d1ee18e65b2325d7.

aceszhenwei committed:
A-Packages 2

Had issues when doing JUnit and had to rollback.

aceszhenwei committed:
A-JUnit

aceszhenwei committed:
A-Jar

aceszhenwei committed:
A-JavaDoc

aceszhenwei committed:
A-CodingStandard

aceszhenwei committed:
Level-9

aceszhenwei committed:
Merge branch 'branch-A-JavaDoc'
Introduces JavaDocs into Duke

aceszhenwei committed:
Merge branch 'master' into branch-A-CodingStandard

aceszhenwei committed:
Merge from branch-A-CodingStandard

A-CodingStandard

aceszhenwei committed:
Update TaskList.java

aceszhenwei committed:
Update TaskList.java

aceszhenwei committed:
Merge branch 'master' into branch-Level-9

aceszhenwei committed:
Merge pull request #2 from aceszhenwei/branch-Level-9

Level-9

aceszhenwei committed:
Merge branch 'master' of https://github.com/aceszhenwei/ip

aceszhenwei committed:
Merge branch 'add-gradle-support'

aceszhenwei committed:
A-Gradle

aceszhenwei committed:
Added Initial GUI Files and Code

Updated build.gradle to prepare for GUI overhaul. Added and coded files as per following the JavaFX tutorial.

aceszhenwei committed:
Fixed JavaFXML Import Bug

Fixed the import bug with help from forum. Additionally, cleaned up certain useless functions due to the transition to FXML.

aceszhenwei committed:
A-Checkstyle

Fixed coding standards via Checkstyle. Note, due to forgetting to start on a different branch, some parts of Level 10 (JavaFX) is now in the master code. Future changes will be continued in a separate branch.

aceszhenwei committed:
Finished Level 10

Current barebones GUI that satisfies the basic requirements of Level 10. Will not merge yet.

aidanaden commented on src/main/java/duke/duke/TaskList.java:
```diff
@@ -0,0 +1,177 @@
+     * @throws IndexOutOfBoundsException If there are no details in input.
+     */
+
+    public void event(String input, ArrayList<Task> arr) {
```
perhaps a better name for the "arr" parameter would be "tasks"?

aidanaden commented on src/main/java/duke/duke/Parser.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    public void parse(String commandString, TaskList tasks, Storage storage) {
+        if (commandString.equals("bye")) {
```
Maybe consider using a switch-case instead? tbh I prefer if else loops too xd

aidanaden commented on src/main/java/duke/duke/Deadline.java:
```diff
@@ -0,0 +1,40 @@
+        this.dueDate = dueDate;
+    }
+
+    /**
```
Not sure if there should be a space between the function and the javadoc comment of the function? based on style guide [here](https://se-education.org/guides/conventions/java/intermediate.html#comments) - think shouldn't have a space between

aidanaden reviewed (commented):
LGTM, just style guide nits to fix! 👍

aceszhenwei committed:
Updated Gradle to Match Launcher

aceszhenwei committed:
Added More JavaDoc

aceszhenwei committed:
Fixed Duke's Packaging

jltham commented on src/main/java/duke/duke/Parser.java:
```diff
@@ -0,0 +1,28 @@
+package duke.duke;
+public class Parser {
+
+    public Parser() {
```
I think this empty constructor is redundant! Maybe you can consider removing it to reduce the LoC of your program?

jltham commented on src/main/java/duke/duke/Deadline.java:
```diff
@@ -0,0 +1,40 @@
+package duke.duke;
```
May want to consider moving the different files into relevant folders? E.g. Tasks and subtasks into a "Tasks" folder, then subsequently package them accordingly for better organization!

jltham commented on src/main/java/duke/duke/Parser.java:
```diff
@@ -0,0 +1,28 @@
+        } else if (commandString.contains("find")) {
+            tasks.find(commandString, tasks.arr);
+        } else {
+            System.out.println(commandString + " is not a valid command. Please try again.");
```
May want to consider throwing a DukeException here as an alternative?

jltham commented on src/main/java/duke/duke/Storage.java:
```diff
@@ -0,0 +1,68 @@
+        ObjectInputStream readStream = new ObjectInputStream(readData);
+        @SuppressWarnings("unchecked")
+        // The object read will always be an ArrayList as it always loads from the previously saved ArrayList.
+        ArrayList<Task> arr = (ArrayList<Task>) readStream.readObject();
```
Although arr is a temporarily declared variable scoped within this method only, you may want to rename it to be more descriptive for better readability? May seem unnecessarily tedious but may prove to be helpful in the future!

jltham commented on src/main/java/duke/duke/TaskList.java:
```diff
@@ -0,0 +1,177 @@
+     * @throws IndexOutOfBoundsException If there are no details in input.
+     */
+
+    public void deadline(String input, ArrayList<Task> arr) {
```
May want to consider renaming to a verb such as "setDeadline" instead of just "deadline"?

jltham reviewed (commented):
Generally looks great! I like how you used different libraries to help you with your code and some abstractions are very clean! 😄 Just a few small things to nitpick on though, but otherwise great job! 🥂

aceszhenwei committed:
GUI Complete

aceszhenwei committed:
Merge pull request #3 from aceszhenwei/branch-Level-10

Merge branch 'branch-Level-10'

aceszhenwei committed:
A-Assertions

aceszhenwei committed:
Fixed mark() exception error

The current mark function only catches NumberFormatExceptions.

This needs to change because OutOfBoundsExceptions may occur and not be caught.

Changed it to catch all Exceptions instead.

This ensures that all Exceptions will be caught and return the correct line instead.

aceszhenwei committed:
Merge pull request #4 from aceszhenwei/branch-A-Assertions

Add Assertions

aceszhenwei committed:
Added Credits

Some of the current codes do not credit authors.

This needs to be changed to fit the Policy of Reuse.

Hence, added all credits to authors.

aceszhenwei committed:
Added Help Feature

Added the help feature to fit iP requirements.

aceszhenwei committed:
Fix Coding Standard

Fixed Coding Standard to fit iP requirements.

In addition, added javadoc folder to .gitignore, to prevent binary files being committed.

aceszhenwei committed:
Merge pull request #6 from aceszhenwei/branch-A-CodeQuality

Fix Coding Standard

aceszhenwei committed:
Ui

Add Ui to match iP requirements.

aceszhenwei committed:
A-UserGuide

aceszhenwei committed:
Update README.md

aceszhenwei committed:
Set theme jekyll-theme-slate

aceszhenwei committed:
Update README.md

aceszhenwei committed:
Update README.md

aceszhenwei committed:
Update README.md
